service-mesh/config.go

```diff
@@ -397,3 +412,129 @@ func downstreamTLSTransportSocket(requireClientCertificate bool) (*envoyCoreV3.T
 		},
 	}, nil
 }
+
+func addBlackHoleToConfig(config *envoyConfigBootstrapV3.Bootstrap) error {
+	// Add blackHoleCluster
+	config.StaticResources.Clusters = append(config.StaticResources.Clusters, blackHoleCluster())
+
+	// Add BlackHole matching to all listeners
+	for _, listener := range config.StaticResources.Listeners {
+		if listener.FilterChainMatcher != nil {
+			return fmt.Errorf("listener %s already has a filterChainMatcher", listener.Name)
+		}
+		listenPort := listener.Address.GetSocketAddress().GetPortValue()
+		if listenPort == 0 {
+			return fmt.Errorf("listener %s listens on port 0", listener.Name)
+		}
+		if listenPort != 15006 && listenPort != 15007 {
+			// listener is none of the ingress listeners
+			continue
+		}
+		if len(listener.FilterChains) != 1 {
+			return fmt.Errorf("listener %s doesn't has exactly one existing listener", listener.Name)
+		}
+		var err error
+		listener.FilterChainMatcher, err = filterChainMatcher(int(listenPort), listener.FilterChains[0].GetName())
+		if err != nil {
+			return fmt.Errorf("could not add filterChainMatcher to listener %s: %w", listener.Name, err)
+		}
+
+		bhFilter, err := blackHoleFilter()
+		if err != nil {
+			return err
+		}
+		listener.FilterChains = append(listener.FilterChains, bhFilter)
+	}
+	return nil
+}
+
+// Blackhole traffic that arrives on the original destination listerners,
+// which original port is the envoy itself, i.e. traffic that was not redirected
+// to the envoy via the TROXY iptables rule. Such traffic would lead to a
+// traffic storm since envoy would connect to the original destination
+// i.e. itself again.
```

Comment:
Please add a sentence that explains how the traffic is blackholed - is it because there is no upstream in the cluster?

Reply:
I hope I've answered your question in the added sentences. If not, please unresolve.